### ISSUE TITLE
(Bug) Wired XInput (Xbox) devices aren't detected

### ISSUE THREAD

**Issue by SkullMan140** [EliteAsian123/YARG#20]
while i can see the "Mouse" and "Keyboard" options, for some reason my Wired Xbox 360 Guitar (Rock Band 1 model) don't appear in the list of devices i can play with while creating a new player profile, even if i use the "refresh" option

**Comment by radobertalan** [comment]
Same with GH3 Xbox 360 Xplorer

I assume since these guitars have native drivers and aren't emulating generic game controllers they'd need separate support. 

but hey, my G29 steering wheel works😅

**Comment by EliteAsian123** [comment]
Thanks for the issue report.

Yes, this issue is known, and I am going to start working on a fix for this now. Unfortunately, I don't have any non-working controllers, so it may take a while to properly diagnose the issue.

Apologies for the inconvinience.

**Comment by EliteAsian123** [comment]
Hello! I've created an updated version of the game that *may* fix the issue. The link for it can be found here:
https://drive.google.com/file/d/1DcKNyK-E-_w1NBM9AgaFBTPP0JFXw9xS/view?usp=share_link

It contains the newest dev-build that has a completely new “Add/Edit Players” menu.

To add an unsupported device:
1. Click “Add/Edit Players” on the main menu.
2. Click “Add Player.”
3. Find the device in the list. It should say “NOT RECOGNIZED” beside it. Click on it.
4. The device should now appear as recognized. Click on it again.
5. Choose input strategy (probably “Five Fret”). Press “Next.”
6. See if you can bind everything. **I have my doubts that this will work**, but it is worth a shot.
7. If you can bind everything, please let me know and have fun! If not, also please let me know. Thanks.

**Comment by AndrewRice2003** [comment]
I've got a rock band 4 wired xbox controller and its not showing up in the device list in game.

**Comment by EliteAsian123** [comment]
> I've got a rock band 4 wired xbox controller and its not showing up in the device list in game.

Is this on the above-mentioned dev-build? (Link here: https://drive.google.com/file/d/1DcKNyK-E-_w1NBM9AgaFBTPP0JFXw9xS/view?usp=share_link)

**Comment by AndrewRice2003** [comment]
Yes, I tried unplugging the controller and checking again to see if it was misreporting the name or something but no. The list was basically just all my other peripherals like mouse, keyboard, headset, and oddly enough my Corsair lighting core.

**Comment by EliteAsian123** [comment]
Okay, thanks for trying it out.

I forced it to list out every peripheral, but the guitar still wasn't there? That means that there is probably an underlying issue.

I did some research, and it seems that [this may be an issue with Unity's input backend](https://forum.unity.com/threads/xbox-one-controller-not-detected-but-works-with-the-older-input-system.813309/). On the forum post I linked, someone stated:

> The Xbox controller is detected and works when connected wirelessly, but not when connected via USB.

And since all the issue reports have something to do with an XInput device hooked-up via USB, I think this is probably the issue. 

So in conclusion: it *probably* isn't my fault. The forum post *says* that the old input system doesn't have this issue, so I may have to switch back to that. The only issue with switching back is that the pro-guitar and others rely on the new input system, so I'll have to come up with some hybrid.

So it may be longer until this issue is fixed. I apologize for that.

Thanks a lot to everyone!

**Comment by TheNathannator** [comment]
This definitely seems like a Unity 2022 issue, I haven't had any issues with XInput devices on 2019.4 or 2021.3 when doing my own antics with the new input system. Maybe it's been fixed in a later 2022 version?

I tried investigating in the editor and neither my wired or wireless guitars are being picked up. However, my Xbox One Elite Series 2 is picked up as a standard XInput gamepad, so Unity's not entirely ignoring XInput devices. Same behavior on the dev build posted here.

**Comment by EliteAsian123** [comment]
> This definitely seems like a Unity 2022 issue, I haven't had any issues with XInput devices on 2019.4 or 2021.3 when doing my own antics with the new input system. Maybe it's been fixed in a later 2022 version?

I hadn't really realized that I've been using a weird beta version until now. I guess that is what I get for not paying attention :/

I'll try to update and see if that fixes the problem. As I've mentioned, I don't have access to any XInput guitars, so it is difficult for me to properly test this.

If that doesn't work, I'll attempt to switch back down to 2021.3 or file a bug report. That being said, the latter of the two will probably not be possible as I can't even reproduce it myself in the first place.

Thanks very much for the help!

**Comment by TheNathannator** [comment]
Yeah I was curious about the version lol

I've updated the project on my end to 2022.11f1 and it seems to still be broken, not seeing anything listed even in the input debugger. I'll go ahead and file a report, rather odd how this got broken.

**Comment by leberj** [comment]
> I've got a rock band 4 wired xbox controller and its not showing up in the device list in game.

Can confirm I'm getting the same issue as this with the third party wireless adapter using the above posted google drive build, all other devices are listed except for the wireless adapter/guitar.

**Comment by EliteAsian123** [comment]
> I've updated the project on my end to 2022.11f1 and it seems to still be broken, not seeing anything listed even in the input debugger. 

That kind of stinks.

> I'll go ahead and file a report, rather odd how this got broken.

Thanks a lot!

> Can confirm I'm getting the same issue as this with the third party wireless adapter using the above posted google drive build, all other devices are listed except for the wireless adapter/guitar.

Thanks for the feedback, and for your initial bug report.

<hr/>

Overall, I'm glad we figured out the root cause. I'm guessing this bug is not being fixed for a while on Unity's end, so I'll try to downgrade to 2021.3 tomorrow and see what happens. I don't think I use any *very* modern features anymore so it should (hopefully) be fine.

Thanks a lot again!

**Comment by TheNathannator** [comment]
Made a post on the Unity forums: https://forum.unity.com/threads/non-gamepad-xbox-360-peripherals-are-not-enumerated-from-the-windows-native-backend-in-unity-2022.1412946/

I should probably also make a bug report through the editor, but I've got other obligations at the moment so I'll do it later lol

**Comment by EliteAsian123** [comment]
> I should probably also make a bug report through the editor, but I've got other obligations at the moment so I'll do it later lol

**Comment by EliteAsian123** [comment]
Hello!

Let's try this one more time. I've downgraded YARG to Unity 2021.3.21f1, so the bug in Unity should be fixed—or rather, not be there yet.

I haven't removed the “UNSUPPORTED” section yet, **please let me know** if your device **does** show up there because I will be removing it.

Here is a link to the (hopefully) fixed version:
https://drive.google.com/file/d/1YPbS7kkHBqSMX9h9YlU0AjcpmZKZqH6x/view?usp=share_link

**Comment by TheNathannator** [comment]
Nice! Works perfectly now, no issues mapping anything.

@radobertalan That seems like Steam may be remapping your controls. By default it maps the Xbox A button to Enter, and Enter will exit out of the mapping menu. Try closing out of Steam or disabling controller configuration for Xbox controllers in its settings.

**Comment by EliteAsian123** [comment]
> Nice! Works perfectly now, no issues mapping anything.

Awesome! I can finally close this issue.

> ...the same thing happens after I exit steam(right click->exit steam), "xbox configuration support" is unchecked under controller settings in steam also.

> By default it maps the Xbox A button to Enter, and it seems Enter will exit out of the mapping menu if you haven't clicked on anything before clicking to map a button 

I'll try to fix this as well.

Thanks very much to everyone for all of their help!

**Comment by EliteAsian123** [comment]
@radobertalan I've moved your bug to a new issue (#23)
